search.py

```diff
@@ -229,43 +284,112 @@ def merge_posting_lists(list1, list2, should_perform_merge_positions = False):
+        if posting.doc_id not in scores:
+            scores[posting.doc_id] = score
+        else:
+            scores[posting.doc_id] += score
+
+    # Now we do the sorting
+    sorted_results = sorted(scores.items(), key=functools.cmp_to_key(comparator))
+
+    return [(doc_id, score) for doc_id, score in sorted_results.items()]
 
-def parse_boolean_query(query):
-    parse_by_arr = query.split(AND_KEYWORD)
-    stemmer = nltk.stem.porter.Stemmer()
-    parse_by_arr = [stemmer.stem(word.lower()) for word in parse_by_arr]
-    if not parse_by_arr:
-        return []
+def parse_boolean_query(terms):
+    """
+    Takes in the array of terms from the query
+    Returns the posting list of all the phrase
+    """
+    # First filter out all the AND keywords from the term array
+    filtered_terms = [term for term in terms if term != AND_KEYWORD]
 
-    first_term = parse_by_arr[0]
+    # Get the posting list of the first word
+    first_term = filtered_terms[0]
     res_posting_list = None
     if " " in first_term:
         res_posting_list = perform_phrase_query(first_term)
     else:
         res_posting_list = find_term(first_term)
 
-    for term in parse_by_arr[1:]:
+    # Do merging for the posting lists of the rest of the terms
+    for term in filtered_terms[1:]:
         term_posting_list = None
         if " " in term:
             term_posting_list = perform_phrase_query(term)
         else:
             term_posting_list = find_term(term)
         res_posting_list = merge_posting_lists(res_posting_list, term_posting_list)
-    return res_posting_list
 
-def parse_free_text_query(query):
+    return get_ranking_for_boolean_query(res_posting_list)
+
+def parse_free_text_query(terms):
+    # TODO: See below (delete once done)
     #Expected to add query expansion, after process(query) is done
     #query = query_expansion(process(query))
-    tokens = tokenize_and_process_query(process(query))
-    res = cosine_score(tokens)
+    res = cosine_score(terms)
     return res
 
+def split_query(query):
+    """
+    split_query extracts out the terms into phrases and terms
+    Assumes that the query is well formed.
+    """
+    start_index = 0
+    is_in_phrase = False
+    is_boolean_query = False
+    current_index = 0
+    terms = []
+
+    while current_index < len(query):
+        current_char = query[current_index]
+        if current_char == "\"":
+            if is_in_phrase:
+                is_in_phrase = False
+                terms.append(query[start_index:current_index])
+                start_index = current_index + 1 # +1 to ignore the space after this
+            else:
+                start_index = current_index + 1
+                is_in_phrase = True
+                is_boolean_query = True
+        elif current_char == " ":
+            # Append the word if not parsing part of phrase
+            if not is_in_phrase:
+                terms.append(query[start_index:current_index])
```

Comment:
It seems that is_boolean_query is set to true when there is phrasal query (which is right since phrasal only occurs in boolean query, confirmed by the thread https://luminus.nus.edu.sg/modules/c8b7b8a5-11c1-470b-bf94-31eb8b05070e/forum/categories/cf28d8ed-c490-4fa2-a5f0-9cc1a8478dec/ce39bf7e-1b14-4ea0-a6bf-615046e7f939 ). 

However, boolean query may still happen without phrasal query. eg tom AND jerry. Hence, it would be good to check if query[start_index:current_index] == AND_KEYWORD and set is_boolean_query to true if it occurs

Reply:
Ah right thanks will fix